consensus/raft/raft.go

```diff
@@ -427,6 +448,147 @@ func (rw *raftWrapper) Peers() ([]string, error) {
+}
+
+//ExistingStateReader does a best effort search for existing raft state files
+//and returns the bytes of the latest raft snapshot
+func ExistingStateReader(cfg *Config) (io.Reader, error){
+	snapPath := "snapshots"
+	stateFilePath := "state.bin"
+
+	if cfg.BaseDir == "" || cfg.DataFolder == "" {
+		return nil, errors.New("Data location not specified in cfg")
+	}
+	snapShotDir := filepath.Join(cfg.BaseDir, cfg.DataFolder, snapPath)
+	
+	// take latest chronological file in the snapshot directory
+	snapFiles, err := ioutil.ReadDir(snapShotDir)
+	if len(snapFiles) == 0 {
+		return nil, errors.New("No snapshots saved")
+	}
+	sort.Sort(byRaftSnapFilename(snapFiles))
+	splitName := strings.Split(snapFiles[0].Name(), "-")
+	if len(splitName) != 3 {
+		return nil, errors.New("No snapshots saved")
+	}
+
+	// look for state file in latest snapshot
+	snapShotStateFile := filepath.Join(snapShotDir, snapFiles[0].Name(), stateFilePath)
+	rawBytes, err := ioutil.ReadFile(snapShotStateFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var state interface {}
+	err = decodeState(rawBytes, &state)
+	if err != nil {
+		return nil, err
+	}
+	stateBytes, err := json.Marshal(state)
+	if err != nil {
+		return nil, err
+	}
+	r := bytes.NewReader(stateBytes)
+	return r, nil
+}
+
+//Reset saves a raft snapshot containing newState to be loaded on restart.
+//Only call when Raft is shutdown.
+func Reset(newState state.State, cfg *Config, raftDataPath string, peers []peer.ID) error{
+	err := cleanupRaft(raftDataPath)
+	if err != nil {
+		return err
+	}
+	snapshotStore, err := hraft.NewFileSnapshotStoreWithLogger(raftDataPath, RaftMaxSnapshots, nil)
+	if err != nil {
+		return err
+	}
+
+	serverAddr := hraft.ServerAddress(peer.IDB58Encode(peers[len(peers) -1]))
+	_, dummyTransport := hraft.NewInmemTransport(serverAddr)
+	var raftSnapVersion hraft.SnapshotVersion
+	raftSnapVersion = 1         // As of v1.0.0 this is always 1                                     
+	raftIndex       := uint64(1) // We reset history to the beginning                                                    
+	raftTerm        := uint64(1) // We reset history to the beginning
+	configIndex     := uint64(1) // We reset history to the beginning
+	srvCfg := MakeServerConf(peers)
+	sink, err := snapshotStore.Create(raftSnapVersion, raftIndex, raftTerm, srvCfg, configIndex, dummyTransport)
+	if err != nil {
+		return err
+	}
+	newStateBytes, err := encodeState(newState)
+	_, err = sink.Write(newStateBytes)
+	if err != nil {
+		return err
+	}
+	err = sink.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func cleanupRaft(raftDataDir string) error {
+        raftDB := filepath.Join(raftDataDir, "raft.db")
+        snapShotDir := filepath.Join(raftDataDir, "snapshots")
+        err := os.Remove(raftDB)
+        if err != nil {
+		return err
+        }
+        err = os.RemoveAll(snapShotDir)
+        return err
+}
+
+func encodeState(state interface{}) ([]byte, error) {
```

Comment:
`encodeState` and `decodeState` should go away (see comment above about using libp2p-raft for this)

Reply:
I agree but am in favor of exporting them from libp2p raft, see my response above.